crates/stages/src/stages/sender_recovery.rs

```diff
@@ -134,7 +142,36 @@ impl<DB: Database> Stage<DB> for SenderRecoveryStage {
         // Iterate over channels and append the sender in the order that they are received.
         for mut channel in channels {
             while let Some(recovered) = channel.recv().await {
-                let (tx_id, sender) = recovered.map_err(|boxed| *boxed)?;
+                let (tx_id, sender) = recovered.map_err(|boxed| match *boxed {
+                    SenderRecoveryStageError::FailedRecovery(err) => {
+                        // we need this so we can report the bad block
+                        let mut tx_block_cursor = match tx.cursor_read::<tables::TransactionBlock>()
+                        {
+                            Ok(cursor) => cursor,
+                            Err(e) => return e.into(),
+                        };
+
+                        // get the block number for the bad transaction
+                        // TODO: remove expect
+                        let block_number =
+                            match tx_block_cursor.seek(err.tx).map(|b| b.map(|(_, bn)| bn)) {
+                                Ok(num) => num,
+                                Err(e) => return e.into(),
+                            }
+                            .expect("tx to be in block");
+
+                        // fetch the sealed header so we can use it in the sender recovery unwind
+                        // TODO: remove expect
+                        let (header, block_hash) = match read_sealed_header(&**tx, block_number) {
+                            Ok(header_hash) => header_hash,
+                            Err(e) => return e.into(),
+                        }
+                        .expect("header to exist for block number");
+
+                        FailedSenderRecoveryError::with_block(header.seal(block_hash))
+                    }
+                    SenderRecoveryStageError::StageError(err) => err,
+                })?;
```

Comment:
same here, not sure why we need all this

Reply:
This is so we can track the block which failed signature validation, since we only check that during the recovery stage

Reply:
removed, re-opening these changes in new PR